app/validators/csv/facilities_validator.rb

```diff
@@ -0,0 +1,107 @@
+class CSV::FacilitiesValidator
+  def self.validate(facilities)
+    new(facilities).validate
+  end
+
+  def initialize(facilities)
+    @facilities = facilities
+    @errors = []
+  end
+
+  attr_reader :errors
+
+  def validate
+    at_least_one_facility
+    duplicate_rows
+    per_facility_validations
+    self
+  end
+
+  private
+
+  STARTING_ROW = 2
+
+  attr_reader :facilities, :organization_name, :facility_group_name
+
+  def at_least_one_facility
+    errors << "Uploaded file doesn't contain any valid facilities" if facilities.blank?
+  end
+
+  def duplicate_rows
+    fields = facilities.map { |facility| facility.slice(:organization_name, :facility_group_name, :name) }
+    errors << "Uploaded file has duplicate facilities" if fields.count != fields.uniq.count
+  end
+
+  def per_facility_validations
+    row_errors = []
+
+    facilities.each.with_index(STARTING_ROW) do |import_facility, row_num|
+      csv_validator = FacilityValidator.new(import_facility.attributes)
+
+      next if csv_validator.valid?
+      next if import_facility.valid? # run model validations
+
+      row_errors << [
+        row_num,
+        import_facility.errors.full_messages.to_sentence,
+        csv_validator.errors.full_messages.to_sentence
+      ]
+    end
+
+    group_row_errors!(row_errors).each { |error| errors << error } if row_errors.present?
+  end
+
+  def group_row_errors!(row_errors)
+    unique_errors = row_errors.map { |_row, message| message }.uniq
+    unique_errors.map do |error|
+      rows = row_errors.select { |row, message| row if error == message }.map { |row, _message| row }
+      "Row(s) #{rows.join(", ")}: #{error}"
+    end
+  end
+
+  class FacilityValidator
+    include ActiveModel::Model
+    include Memery
+
+    validates :name, presence: true
+    validates :organization_name, presence: true
+    validates :facility_group_name, presence: true
+    validate :facility_is_unique
+    validate :organization_exists
+    validate :facility_group_exists
+
+    attr_accessor(*Facility.new.attributes.keys.map(&:to_sym))
```

Comment:
Do we need accessors for all the attributes, looks like just `:name` will do?

Reply:
True, I just became lazy, will fix.